### DISCUSSION TITLE
Non Blocking Future check

### DISCUSSION THREAD

**Discussion by ghost** [jodal/pykka#69] [Q&A]
I've implemented now a "pool of resolver" with:

``` python
def _scan(uri, timeout=1):
    pass
     # do something

def _future_check(uri_to_check, timeout=1):
    urilist = pykka.ThreadingFuture()       # create a future object
    results = urilist.map(partial(_scan, timeout=timeout))

    # start elaboration
    urilist.set(uri_to_check)

    # attention: results are future so they need to be retrieved using get()
   return results
```

how i can check if all futures are arrived without blocking???
there's a "look into future" function??

``` python
def future_is_ready(future, timeout=0.05):
    try:
        future.get(timeout=timeout)
        return True
    except pykka.Timeout:
        return False


results = _future_check(uris)
if future_is_ready(results):
    print(results.get())

```

**Comment by adamcik** [comment]
Add least for the threading version you have https://github.com/jodal/pykka/blob/develop/pykka/threading.py#L50 and you would want to call the underlying queue with `False` so it becomes non-blocking. So I don't think there is a way of doing this cleanly without adding a block argument to `get`.

What could work is a `get_all` with a very short, but non-negative timeout. Then you would just check catch the timeout exception, in which case something isn't done. Otherwise you are all good.

**Comment by ghost** [comment]
tornado for example

http://www.tornadoweb.org/en/stable/concurrent.html

have two mechanism
the add_done_callback()
that call a function when the future is done
and a done() function to check if future is done...

https://github.com/tornadoweb/tornado/blob/master/tornado/concurrent.py
